File: java/solutions/src/test/java/io/github/ajoz/workshop/fp/part3/exercise1/EnumBooleanTest.java
package io.github.ajoz.workshop.fp.part3.exercise1;

import org.junit.Assert;
import org.junit.Test;
import org.junit.experimental.runners.Enclosed;
import org.junit.runner.RunWith;

import static org.junit.Assert.assertEquals;

@RunWith(Enclosed.class)
public class EnumBooleanTest {
    public static class True {
        @Test
        public void and_True_equals_True() {
            Assert.assertEquals(EnumBoolean.TRUE, EnumBoolean.TRUE.and(EnumBoolean.TRUE));
        }

        @Test
        public void and_False_equals_False() {
            Assert.assertEquals(EnumBoolean.FALSE, EnumBoolean.TRUE.and(EnumBoolean.FALSE));
        }

        @Test
        public void or_True_equals_True() {
            Assert.assertEquals(EnumBoolean.TRUE, EnumBoolean.TRUE.or(EnumBoolean.TRUE));
        }

        @Test
        public void or_False_equals_True() {
            Assert.assertEquals(EnumBoolean.TRUE, EnumBoolean.TRUE.or(EnumBoolean.FALSE));
        }

        @Test
        public void negated_equals_False() {
            Assert.assertEquals(EnumBoolean.FALSE, EnumBoolean.TRUE.not());
        }
    }

    public static class False {
        @Test
        public void and_True_equals_False() {
            Assert.assertEquals(EnumBoolean.FALSE, EnumBoolean.FALSE.and(EnumBoolean.TRUE));
        }

        @Test
        public void and_False_equals_False() {
            Assert.assertEquals(EnumBoolean.FALSE, EnumBoolean.FALSE.and(EnumBoolean.FALSE));
        }

        @Test
        public void or_True_equals_True() {
            Assert.assertEquals(EnumBoolean.TRUE, EnumBoolean.FALSE.or(EnumBoolean.TRUE));
        }

        @Test
        public void or_False_equals_True() {
            Assert.assertEquals(EnumBoolean.FALSE, EnumBoolean.FALSE.or(EnumBoolean.FALSE));
        }

        @Test
        public void negated_equals_True() {
            Assert.assertEquals(EnumBoolean.TRUE, EnumBoolean.FALSE.not());
        }
    }
}

File: kotlin/exercises/src/test/kotlin/io/github/ajoz/workshop/fp/kotlin/part_3/exercises/exercise_1/EnumBooleanTest.kt
@file:Suppress("PackageName")

package io.github.ajoz.workshop.fp.kotlin.part_3.exercises.exercise_1

import io.github.ajoz.workshop.fp.kotlin.part_3.exercises.exercise_1.EnumBoolean.FALSE
import io.github.ajoz.workshop.fp.kotlin.part_3.exercises.exercise_1.EnumBoolean.TRUE
import org.junit.Test
import org.junit.experimental.runners.Enclosed
import org.junit.runner.RunWith

import org.junit.Assert.assertEquals

@RunWith(Enclosed::class)
class EnumBooleanTest {
    class True {
        @Test
        fun `and True equals True`() {
            assertEquals(TRUE, TRUE and TRUE)
        }

        @Test
        fun `and False equals False`() {
            assertEquals(FALSE, TRUE and FALSE)
        }

        @Test
        fun `or True equals True`() {
            assertEquals(TRUE, TRUE or TRUE)
        }

        @Test
        fun `or False equals True`() {
            assertEquals(TRUE, TRUE or FALSE)
        }

        @Test
        fun `negated equals False`() {
            assertEquals(FALSE, TRUE.not())
        }
    }

    class False {
        @Test
        fun `and True equals False`() {
            assertEquals(FALSE, FALSE and TRUE)
        }

        @Test
        fun `and False equals False`() {
            assertEquals(FALSE, FALSE and FALSE)
        }

        @Test
        fun `or True equals True`() {
            assertEquals(TRUE, FALSE or TRUE)
        }

        @Test
        fun `or False equals True`() {
            assertEquals(FALSE, FALSE or FALSE)
        }

        @Test
        fun `negated equals True`() {
            assertEquals(TRUE, FALSE.not())
        }
    }
}